feat: add startsAt, endsAt, and duration fields to v1alpha2 Silence CRD

QuentinBisson opened the pull request:
Adds explicit silence scheduling to the `observability.giantswarm.io/v1alpha2` Silence CRD.

## Changes

**New spec fields:**
- `startsAt` — RFC 3339 timestamp for when the silence becomes active (defaults to creation time)
- `endsAt` — RFC 3339 timestamp for when the silence expires
- `duration` — how long the silence is active from `startsAt`; accepts weeks (`w`), days (`d`), hours (`h`), minutes (`m`), seconds (`s`): `"7d"`, `"2w"`, `"1d12h"`, `"30m"`
- `endsAt` and `duration` are mutually exclusive (CEL validation); `startsAt` must precede `endsAt`

**Priority chain** when computing end time:
1. `spec.endsAt`
2. `spec.startsAt` + `spec.duration`
3. `valid-until` annotation (migration path from v1alpha1)
4. 100-year default (v1alpha1 backward compatibility)

**Domain type:** `SilenceDuration string` in `api/v1alpha2` with a `Duration()` method that expands `d`/`w` to hours before delegating to `time.ParseDuration`. The controller calls `silence.Spec.Duration.Duration()` directly.

**CRD validation:** `pattern: ^(\d+(w|d|h|m|s))+$` on the `duration` field.

**Tests:**
- Table-driven unit tests for `SilenceDuration.Duration()` covering all unit combinations and invalid inputs
- Integration tests for: `endsAt` priority, `duration` with explicit `startsAt`, `duration` with creation timestamp, `valid-until` annotation fallback, CEL rejection of both fields set simultaneously, matcher type conversion, finalizer lifecycle

**Docs:** README, MIGRATION.md, CHANGELOG, and samples updated.

Closes #580

QuentinBisson committed:
**Enhanced v1alpha2 Silence CRD with flexible time management**

Adds `startsAt`, `endsAt`, and `duration` fields with priority-based resolution, comprehensive testing, and refactored matcher conversion while maintaining full backward compatibility.

QuentinBisson committed:
Fix review findings: free convertMatchers, correct duration syntax, clean up tests and docs

QuentinBisson committed:
Introduce SilenceDuration type with week/day support

- Define `SilenceDuration string` in api/v1alpha2 with a `Duration()` method
  that expands d→24h and w→168h before delegating to time.ParseDuration.
- Change `SilenceSpec.Duration` from `*metav1.Duration` to `*SilenceDuration`
  so CRD serialization remains a plain string while the domain type owns parsing.
- Remove parseSilenceDuration and durationExtPattern from the controller; the
  Duration branch now calls silence.Spec.Duration.Duration().
- Update zz_generated.deepcopy.go to use new(SilenceDuration).
- Add table-driven unit tests covering plain units (h, m, s), extended units
  (d, w), combinations (1d12h, 2w3d), and invalid inputs.
- Update CRD YAML (config/ and helm/) with pattern ^(\d+(w|d|h|m|s))+$ and
  corrected duration description.
- Update integration tests to use SilenceDuration string literals.
- Update README, MIGRATION.md, CHANGELOG, and samples to reflect the new
  duration syntax and valid-until fallback chain.

QuentinBisson committed:
fix: address review findings in duration/time scheduling code

- Hoist durationMultipliers to package-level var (was re-allocated per regex match)
- Switch Duration() error to fmt.Errorf; pkg/errors stack traces are inappropriate in an API types package
- Replace time.Now() fallback with an error return; non-deterministic endsAt when creationTimestamp is zero hides a bug rather than surfacing it
- Fix misleading comment: valid-until annotation is an absolute timestamp, not added to startsAt
- Remove verbose convertMatchers function comment
- Trim redundant mock alertmanager comments

QuentinBisson committed:
Merge branch 'main' into enhance-time-management

QuentinBisson committed:
fix: set CreationTimestamp in MatchType conversion test

TheoBrigitte commented on api/v1alpha2/silence_types.go:
```diff
@@ -35,6 +40,28 @@ const (
+	}
+	return d, nil
+}
+
```
This works, but I'd rather we rely on some already existing library https://github.com/xhit/go-str2duration.

TheoBrigitte commented on api/v1alpha2/silence_types.go:
```diff
@@ -58,12 +85,32 @@ type SilenceSpec struct {
+	EndsAt *metav1.Time `json:"endsAt,omitempty"`
+
+	// Duration defines how long the silence is active from StartsAt (or creation time when StartsAt is unset).
+	// Ignored when EndsAt is set. When neither EndsAt nor Duration is set, the valid-until annotation is used,
```
```suggestion
	// Ignored when EndsAt is set. Takes precedence over the valid-until annotation.
```

TheoBrigitte commented on config/crd/bases/observability.giantswarm.io_silences.yaml:
```diff
@@ -43,6 +52,19 @@ spec:
+              duration:
+                description: |-
+                  Duration defines how long the silence is active from StartsAt (or creation time when StartsAt is unset).
+                  Ignored when EndsAt is set. When neither EndsAt nor Duration is set, the valid-until annotation is used,
```
```suggestion
                  Ignored when EndsAt is set. Ignored when EndsAt is set. Takes precedence over the valid-until annotation.
```

TheoBrigitte commented on helm/silence-operator/templates/crds.yml:
```diff
@@ -142,6 +151,19 @@ spec:
+              duration:
+                description: |-
+                  Duration defines how long the silence is active from StartsAt (or creation time when StartsAt is unset).
+                  Ignored when EndsAt is set. When neither EndsAt nor Duration is set, the valid-until annotation is used,
```
```suggestion
                  Ignored when EndsAt is set. Ignored when EndsAt is set. Takes precedence over the valid-until annotation.
```

QuentinBisson commented:
@TheoBrigitte all four comments addressed in 0745dc3.

`SilenceDuration.Duration()` now delegates to `str2duration.ParseDuration`, so the regex expansion and the multiplier map are gone. The library was already an indirect dependency via the Alertmanager side, so `go mod tidy` only promoted it to direct. Behaviour is identical, I checked it against every case in the existing table test before swapping.

Doc wording fixed in the Go comment and both CRD copies regenerated (`make manifests` + `make update-helm-crds`). Note your two YAML suggestions had "Ignored when EndsAt is set." twice, I kept the single sentence.

Also tightened the duration pattern in 4c940d9. The old `^(\d+(w|d|h|m|s))+$` accepted out-of-order and repeated units, so `1h30d` and `2d2d` were silently parsed as 721h and 96h instead of being rejected. Units now have to appear at most once, ordered largest to smallest. Since the field is still unreleased this costs no migration, whereas after the next tag it would be a breaking API change. Added envtest cases for the rejected forms, and updated README, MIGRATION and CHANGELOG.

TheoBrigitte reviewed (approved):
LGTM
